facebook/docusaurus
# fix: rename version headers recursively and added copy-version command

## Motivation

To align docusaurus' versioning system with a production pipeline, you sometimes need to make a copy of a version. Copying is different than renaming as you keep track of the version like a branch has a remote in Github.

### Have you read the [Contributing Guidelines on pull requests]
No, but feel free to add any comment.

This PR fixes the existing `rename-version`. `rename-version` correctly creating the file structure of the version and changing the headers of the top level markdown files. However, if you `docs` folder was containing sub-folders, the header are currently not updated.

This fix modifies the headers recursively down the tree. To make thing reusable, a new package `headerUtils` has been created under `/lib/server` (please advise if there is a better location)

It also introduce a `docusaurus-copy-server` command that allows to create a version from an existing one. This is pretty interesting when having multiple environment as your "latest" might not be what is deployed in your production or preproduction environment.

## Test Plan

1) Create a markdown file under `docs`
2) Create a subfolder containing a markdown file under `docs`
3) Create a version `production` using the regular command:

 ```
yarn run version production
```

> Expectations
> A) Expect all the markdown files under `versioned_docs/version-production` to have a header containing `id: version-production-xxx`

4) Copy the `production` version into the `v0.0.1` as it is now official:

```
yarn run copy-version production v0.0.1
```

> Expectations
> A) Expect all the markdown files under `versioned_docs/version-production` to have a header containing `id: version-production-xxx`
> B) Expect all the markdown files under `versioned_docs/version-v0.0.1` to have a header containing `id: version-v0.0.1-xxx`

5) Rename version `v0.0.1` to `v1.0.0` as you have made a mistake.

```
yarn run rename-version v0.0.1 v1.0.0
```

> Expectations
> A) Expect all the markdown files under `versioned_docs/version-production` to have a header containing `id: version-production-xxx`
> B) Expect all the markdown files under `versioned_docs/version-v0.0.1` to have a header containing `id: version-v0.0.1-xxx`
> C) Expect all the markdown files under `versioned_docs/version-v1.0.0` to have a header containing `id: version-v1.0.0-xxx`

## Related PRs

No related PRs.
